### PR TITLE
[SYCLomatic] Enable migration for CUBLASLT_EPILOGUE_DGELU & EPILOGUE_BGRADB

### PR DESCRIPTION
**Pull request by abhilash1910**
@zhiweij1 @zhimingwang36 
(WIP)

### CODE REVIEW

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 1]
```diff
@@ -44,6 +44,19 @@ using matmul_desc_ptr = matmul_desc_t *;
 class transform_desc_t;
 using transform_desc_ptr = transform_desc_t *;
 
+template <typename primitive_type, typename... args_type>
+inline
+typename primitive_type::primitive_desc dgelu_epilogue_(
+    float alpha, float beta) {
+
+  auto alg = ::dnnl::algorithm::eltwise_gelu_erf;
+  const memory_desc_ext &dst_desc = new memory_desc_ext();
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: where to "delete" `dst_desc` and `src_desc`?

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 2]
```diff
@@ -44,6 +44,19 @@ using matmul_desc_ptr = matmul_desc_t *;
 class transform_desc_t;
 using transform_desc_ptr = transform_desc_t *;
 
+template <typename primitive_type, typename... args_type>
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: If is only for internal usage, please move it into `namespace detail`

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 3]
```diff
@@ -44,6 +44,19 @@ using matmul_desc_ptr = matmul_desc_t *;
...
+inline
+typename primitive_type::primitive_desc dgelu_epilogue_(
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: The function name looks like a type name, please refine it.

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 4]
```diff
@@ -44,6 +44,19 @@ using matmul_desc_ptr = matmul_desc_t *;
...
+    float alpha, float beta) {
+
+  auto alg = ::dnnl::algorithm::eltwise_gelu_erf;
+  const memory_desc_ext &dst_desc = new memory_desc_ext();
+  const memory_desc_ext &src_desc = new memory_desc_ext();
+  return create_primitive_desc<primitive_type>(
+                    ::dnnl::prop_kind::backward, alg, src_desc.get_desc(),
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: For `src_desc` and `dst_desc`, is it OK to pass to the `create_primitive_desc` directly?
No need to fill data?

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 5]
```diff
@@ -99,8 +104,6 @@ class matrix_layout_t {
       CASE(rows)
       CASE(cols)
       CASE(ld)
-      CASE(batch_count)
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: why remove it?

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 6]
```diff
@@ -62,15 +75,7 @@ class descriptor {
 
 class matrix_layout_t {
 public:
-  enum class attribute {
-    type,
-    order,
-    rows,
-    cols,
-    ld,
-    batch_count,
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: why remove it?

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 7]
```diff
@@ -110,8 +113,6 @@ class matrix_layout_t {
   std::uint64_t _rows;
   std::uint64_t _cols;
   std::int64_t _ld;
-  std::uint64_t _batch_count = 1;
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: why remove it?

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 8]
```diff
@@ -137,8 +138,7 @@ class matmul_desc_t {
     trans_b,
     trans_c,
     epilogue,
-    epilogue_aux_ld,
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: why remove it?

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 9]
```diff
@@ -180,8 +180,7 @@ class matmul_desc_t {
       CASE(b_scale_pointer)
       CASE(d_scale_pointer)
       CASE(absmax_d_pointer)
-      CASE(epilogue_aux_ld)
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp:188]: why remove it?

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 10]
```diff
@@ -195,12 +194,12 @@ class matmul_desc_t {
   oneapi::mkl::transpose _trans_b = oneapi::mkl::transpose::nontrans;
   oneapi::mkl::transpose _trans_c = oneapi::mkl::transpose::nontrans;
   epilogue_t _epilogue = epilogue_t::nop;
-  size_t _epilogue_aux_ld = 0;
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: why remove it?

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 11]
```diff
@@ -195,12 +194,12 @@ class matmul_desc_t {
...
   void *_a_scale_pointer = nullptr;
   void *_b_scale_pointer = nullptr;
   void *_d_scale_pointer = nullptr;
   void *_absmax_d_pointer = nullptr;
-  void *_epilogue_aux_pointer = nullptr;
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp:212]: why remove it?

**Review by zhiweij1 — CHANGES_REQUESTED** [review]
Please refine the PR.

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp (hunk 11 above)
Comment (abhilash1910) [reply clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp:212]: let me rebase.

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp (hunk 3 above)
Comment (abhilash1910) [reply clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: OK

**Review by zhiweij1 — COMMENTED** [review]
Please submit E2E test PR first.

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 12]
```diff
@@ -222,6 +229,31 @@ class matmul_desc_t {
...
+inline
+typename primitive_type::primitive_desc bias_backward(
+    float alpha, float beta, const dnnl::memory_desc_ext &src_desc,
+    const ::dnnl::memory_desc_ext &dest_desc) {
+
+  auto alg = ::dnnl::algorithm::reduction_sum;
+  return create_primitive_desc<primitive_type>(
+                    ::dnnl::prop_kind::backward, alg, src_desc.get_desc(),
+                    dst_desc.get_desc(), alpha, beta);
+}
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp:255]: For backward primitives, it needs a forward primitive object as a constructor argument. https://github.com/oneapi-src/oneDNN/blob/main/include/oneapi/dnnl/dnnl.hpp#L7476

**Comment by abhilash1910** [comment]
Transfer to @zhiweij1 . Thanks

**Pull request closed without merging**